src/shared/uri.ts

```diff
@@ -0,0 +1,142 @@
+import * as sourcegraph from 'sourcegraph'
+import { URL as _URL } from 'url'
+
+declare class URL extends _URL {
+    constructor(urlStr: string)
+}
+
+/**
+ * A serialized representation of a {@link URI} produced by {@link URI#toJSON} and that can be deserialized with
+ * {@link URI.fromJSON}.
+ */
+export interface URIComponents {
+    scheme: string
+    authority: string
+    path: string
+    query: string
+    fragment: string
+}
+
+/**
+ * A uniform resource identifier (URI), as defined in [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3).
+ *
+ * This URI class should be used instead of WHATWG URL because it parses all URI schemes in the same way on all
+ * platforms. The WHATWG URL spec requires "non-special schemes" (anything other than http, https, and a few other
+ * common schemes; see https://url.spec.whatwg.org/#special-scheme) to be parsed differently in a way that is not
+ * desirable. For example:
+ *
+ * With WHATWG URL:
+ *
+ *   new URL("https://foo/bar").pathname === "/bar"
+ *
+ *   new URL("git://foo/bar").pathname === "//foo/bar" // UNDESIRABLE
```

Comment:
This is not how WHATWG URL behaves: https://runkit.com/felixfbecker/5be0d65fa5064900123995f4

Reply:
Then I think it's not spec compliant. Try it in the browser and read the https://url.spec.whatwg.org/#special-scheme doc.